performance/class-media-library-caching.php

```diff
@@ -0,0 +1,174 @@
+	public const AVAILABLE_MIME_TYPES_CACHE_KEY     = 'vip_available_mime_types';
+	public const USING_DEFAULT_MIME_TYPES_CACHE_KEY = 'vip_using_default_mime_types';
+	public const MAX_POSTS_TO_QUERY_DEFAULT         = 500000;
+
+	/**
+	 * Class initialization.
+	 */
+	public static function init() {
+		global $wp_version;
+
+		/**
+		 * Filters whether the MIME type caching is enabled.
+		 *
+		 * @param bool $cache_mime_types Whether the MIME type caching is enabled. Default false.
+		 */
+		$vip_cache_mime_types = apply_filters( 'vip_cache_mime_types', true );
+
+		if ( $vip_cache_mime_types &&
+			isset( $wp_version ) &&
+			version_compare( $wp_version, self::MINIMUM_WORDPRESS_VERSION, '>=' ) ) {
+			self::enable_post_mime_types_caching();
+		}
+	}
+
+	/**
+	 * Enable MIME type caching.
+	 */
+	private static function enable_post_mime_types_caching() {
+		add_filter( 'pre_get_available_post_mime_types', array( __CLASS__, 'get_cached_post_mime_types' ), 10, 2 );
+		add_action( 'add_attachment', array( __CLASS__, 'update_post_mime_types_cache_on_add' ), 10, 1 );
+		add_action( 'attachment_updated', array( __CLASS__, 'update_post_mime_types_cache_on_edit' ), 10, 3 );
+		add_action( 'delete_attachment', array( __CLASS__, 'update_post_mime_types_cache_on_delete' ), 10 );
+	}
+
+	/**
+	 * Get cached results for get_available_post_mime_types() to avoid a query on every page load.
+	 *
+	 * @param string[]|null $filtered_mime_types An array of MIME types. Default null.
+	 * @param string        $type                The post type name.
+	 * @return array An array of MIME types.
+	 */
+	public static function get_cached_post_mime_types( $filtered_mime_types, $type ) {
+		if ( 'attachment' !== $type ) {
+			return $filtered_mime_types;
+		}
+
+		$mime_types = wp_cache_get( self::AVAILABLE_MIME_TYPES_CACHE_KEY, self::CACHE_GROUP );
+
+		if ( false === $mime_types ) {
+
+			/**
+			 * Filters the max number of posts to query for dynamic MIME type caching.
+			 *
+			 * @param bool $max_posts_to_query Max number of posts to query.
+			 */
+			$max_posts_to_query = apply_filters( 'vip_max_posts_to_query_for_mime_type_caching', self::MAX_POSTS_TO_QUERY_DEFAULT );
+
+			global $wpdb;
+			$attachment_count = $max_posts_to_query > 0 ? $wpdb->get_var( $wpdb->prepare( "SELECT COUNT(1) FROM {$wpdb->posts} WHERE post_type = %s", $type ) ) : 1;
+			$use_defaults     = $attachment_count > $max_posts_to_query;
+
+			if ( $use_defaults ) {
+				// If there are too many posts to query, use the default MIME types.
+				$mime_types = self::get_default_mime_types();
+			} else {
+				// Otherwise, use the same query from core.
+				$mime_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s", $type ) );
+			}
+
+			// Cache the results.
+			wp_cache_set( self::AVAILABLE_MIME_TYPES_CACHE_KEY, $mime_types, self::CACHE_GROUP );
+			wp_cache_set( self::USING_DEFAULT_MIME_TYPES_CACHE_KEY, $use_defaults, self::CACHE_GROUP );
```

Comment:
May end up in a confusing scenario where one gets evicted but the other is not. I'd prefer we just keep this data all inside one cache key I think, better for slight performance gains as well. The cache value can just be like:

```
[
 'available_mime_types' => [...]
 'using_default_mime_types' => true|false
]
```

Then we just fetch one cache key and can check for the specific values when needed.

Reply:
Went with a `vip_mime_types` cache key, which contains an array with the following keys (to avoid name redundancy):
```
[
 'available_types' => [...]
 'using_defaults' => true|false
]
```
And refactored everything to handle a single cache entry.